pkg/controller/config/config_controller_test.go

```diff
@@ -196,6 +199,147 @@ func TestReconcile(t *testing.T) {
 	cs.Stop()
 }
 
+// tests that expectations for sync only resource gets cancelled when it gets deleted
+func TestConfig_DeleteSyncResources(t *testing.T) {
+	log.Info("Running test: Cancel the expectations when sync only resource gets deleted")
+
+	g := gomega.NewGomegaWithT(t)
+
+	// setup the Manager and Controller.  Wrap the Controller Reconcile function so it writes each request to a
+	// channel when it is finished.
+	mgr, wm := setupManager(t)
+	c := mgr.GetClient()
+
+	// create the Config object and expect the Reconcile to be created when controller starts
+	instance := &configv1alpha1.Config{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:       "config",
+			Namespace:  "gatekeeper-system",
+			Finalizers: []string{finalizerName},
+		},
+		Spec: configv1alpha1.ConfigSpec{
+			Sync: configv1alpha1.Sync{
+				SyncOnly: []configv1alpha1.SyncOnlyEntry{
+					{Group: "", Version: "v1", Kind: "Pod"},
+				},
+			},
+		},
+	}
+	err := c.Create(context.TODO(), instance)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+
+	defer func() {
+		err = c.Delete(context.TODO(), instance)
+		g.Expect(err).NotTo(gomega.HaveOccurred())
+	}()
+
+	// create the pod that is a sync only resource in config obj
+	pod := &corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Pod",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "testpod",
+			Namespace: "default",
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name:  "nginx",
+					Image: "nginx",
+				},
+			},
+		},
+	}
+	g.Expect(c.Create(context.TODO(), pod)).NotTo(gomega.HaveOccurred())
+
+	// set up tracker
+	tracker, err := readiness.SetupTracker(mgr)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+
+	// events channel will be used to receive events from dynamic watches
+	events := make(chan event.GenericEvent, 1024)
+
+	// set up controller and add it to the manager
+	err = setupController(mgr, wm, tracker, events)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+
+	// start manager that will start tracker and controller
+	stopMgr, mgrStopped := StartTestManager(mgr, g)
+	once := gosync.Once{}
+	defer func() {
+		once.Do(func() {
+			close(stopMgr)
+			mgrStopped.Wait()
+		})
+	}()
+	gvk := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Pod"}
+
+	// get the object tracker for the synconly pod resource
+	tr, ok := tracker.ForData(gvk).(testExpectations)
+	if !ok {
+		t.Fatalf("unexpected tracker, got %T", tr)
+	}
+
+	// ensure that expectations are set for the constraint gvk
+	g.Eventually(func() bool {
+		return tr.ExpectedContains(gvk, types.NamespacedName{Name: "testpod", Namespace: "default"})
+	}, timeout).Should(gomega.BeTrue())
+
+	// register events for the pod to go in the event channel
+	podObj := &corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Pod",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "testpod",
+			Namespace: "default",
+		},
+	}
+	events <- event.GenericEvent{
```

Comment:
Should this event be sent _after_ the pod is deleted?

Reply:
this is a generic event channel and any event for the pod will be sent to this channel, so when we delete the pod the delete event is sent to this channel, hence ordering doesn't matter.

Reply:
Thanks Oren, its a good catch, it would be safe to send the event after deleting and would prevent race condition b/w update and delete of the resource. updated the PR.